services/opentsdb/service.go

```diff
@@ -297,20 +341,48 @@ func (s *Service) Addr() net.Addr {
 }
 
 // serve serves the handler from the listener.
-func (s *Service) serve() {
-	for {
-		// Wait for next connection.
-		conn, err := s.ln.Accept()
-		if opErr, ok := err.(*net.OpError); ok && !opErr.Temporary() {
-			s.Logger.Info("OpenTSDB TCP listener closed")
-			return
-		} else if err != nil {
-			s.Logger.Info("Error accepting OpenTSDB", zap.Error(err))
-			continue
+func (s *Service) serve(ctx context.Context) error {
+	conChan := make(chan net.Conn)
+	errChan := make(chan error)
+
+	// send accepted connections to our conChan so that we can multiplex over
+	// conChand and ctx.Done() in the next for loop.
+	//
+	// this go routine is cancelled once the s.ln listener is closed.
+	//
+	// FIXME: this logic needs to be looked at closely.
+	go func() {
+		for {
+			// Wait for next connection.
+			c, err := s.ln.Accept()
+			if opErr, ok := err.(*net.OpError); ok && !opErr.Temporary() {
+				s.Logger.Info("OpenTSDB TCP listener closed")
+				errChan <- err
+				return
+			} else if err != nil {
+				s.Logger.Info("Error accepting OpenTSDB", zap.Error(err))
+				continue
+			}
+			conChan <- c
 		}
+	}()
```

Comment:
My understanding for moving the listener to a separate goroutine and forwarding the new connections over a channel is so that you can read the `context.Context` `Done` channel for cancellation. 

Perhaps in a future update, we could simplify this code by changing `RunWithReady` and build the `Listener` instances with `context.Context`:

```go
var lc ListenConfig
s.ln = lc.Listen(ctx, "tcp", s.BindAddress)
```

as `net.Listen` is:

```go
func Listen(network, address string) (Listener, error) {
	var lc ListenConfig
	return lc.Listen(context.Background(), network, address)
}
```